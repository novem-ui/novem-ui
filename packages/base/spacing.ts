export { default as spacing } from './src/spacing'
export { SpacingProps } from './src/types'
