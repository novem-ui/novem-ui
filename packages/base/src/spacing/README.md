# spacing

Spacing in Novem UI is inspired on well known frontend frameworks/libs such as Chakra UI or Tailwind.

The properties are `margin` (`m`), `marginX` (`mx`), `marginY` (`my`), `marginTop` (`mt`), `marginRight` (`mr`), `marginBottom` (`mb`), `marginLeft` (`ml`), `padding` (`p`), `paddingX`(`py`), `paddingY`(`py`), `paddingTop` (`pt`), `paddingRight` (`pr`), `paddingBottom` (`pb`), `paddingLeft` (`pl`), the values for those properties are `0, 0.5, 1, 1.2, 1.5, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 20, 24, 28, 32, 36, 40, 44, 48, 52, 56, 60, 64, 72, 80, 96` (always as numbers, anything setted as a string will be treated as a normal css value).

## Example

This component

```tsx
<Box pt={4} paddingX="2px" />
```

will produce this css

```css
padding-top: 1rem;
padding-left 2px;
padding-right: 2px;
```