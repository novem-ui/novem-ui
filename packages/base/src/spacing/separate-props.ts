import { SpacingProps } from '../types'

const separateSpacingProps = <T = Record<string, any>>(props: SpacingProps & T) => {
  const {
    m,
    mt,
    mr,
    mb,
    ml,
    mx,
    my,
    margin,
    marginX,
    marginY,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    p,
    pt,
    pr,
    pb,
    pl,
    px,
    py,
    padding,
    paddingX,
    paddingY,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft,
    ...restProps
  } = props

  const spacingProps: SpacingProps = {
    m,
    mt,
    mr,
    mb,
    ml,
    mx,
    my,
    margin,
    marginX,
    marginY,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    p,
    pt,
    pr,
    pb,
    pl,
    px,
    py,
    padding,
    paddingX,
    paddingY,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft
  }

  const typedSpacingProps: Omit<T, keyof SpacingProps> = restProps

  return { spacingProps, props: typedSpacingProps }
}

export default separateSpacingProps
