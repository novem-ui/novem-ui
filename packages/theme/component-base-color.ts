export { ComponentBaseColorProps, ComponentBaseColorProvider, useComponentBaseColor } from './src/component-base-color'
