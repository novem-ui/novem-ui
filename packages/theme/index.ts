export { default as useTheme } from './use-theme'
export { default as withTheme } from './with-theme'
export { default as createColorMap } from './create-color-map'
export { default as ThemeProvider } from './theme-provider'
export { default, Theme, BaseColor, Color, HeadingSizeValue, SpacingValue, TextSizeValue, Shadows } from './theme'
