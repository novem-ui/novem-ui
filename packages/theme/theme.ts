export { default, Theme, BaseColor, Color, HeadingSizeValue, SpacingValue, TextSizeValue } from './src/theme'
