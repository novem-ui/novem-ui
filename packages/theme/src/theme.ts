import colors from './colors'
import * as typography from './typography'
import spacing from './spacing'
import { WeightedColorKey } from '../types'

const theme = {
  colors,
  typography,
  spacing,
}

export type BaseColor = `${keyof typeof colors.palette}`
export type Color = `${keyof typeof colors.palette}.${WeightedColorKey}`
export type HeadingSizeValue = keyof typeof typography.headingFontSizes
export type SpacingValue = keyof typeof spacing
export type TextSizeValue = keyof typeof typography.textFontSizes
export type Theme = typeof theme


export default theme
