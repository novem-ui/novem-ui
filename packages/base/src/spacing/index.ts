import { css } from '@emotion/react'
import { Theme } from '@novem-ui/theme'

import SpacingUtils from '../utils/spacing-utils'
import { CSSSpacingProps, SpacingPropValue } from '../types'

const spacing = ({ theme, ...spacingProps }: { theme?: Theme } & CSSSpacingProps) => {
  const {
    margin,
    marginX,
    marginY,
    marginBottom,
    marginLeft,
    marginRight,
    marginTop,
    padding,
    paddingX,
    paddingY,
    paddingBottom,
    paddingLeft,
    paddingRight,
    paddingTop
  } = SpacingUtils.validateProps(spacingProps)
  const getPropertyValue = (value: SpacingPropValue) => {
    const isString = typeof value === 'string'

    if (isString) {
      return value
    }

    const isNumber = typeof value === 'number'

    if (isNumber) {
      return `${theme.spacing[value]}rem`
    }

    throw new TypeError(`value argument expected a number or a value expressed in CSS units but got ${value} instead`)
  }

  const style = css`
    ${margin && `margin: ${getPropertyValue(margin)};`}
    ${marginX && `margin-left: ${getPropertyValue(marginX)};\nmargin-right:${getPropertyValue(marginX)};`}
    ${marginY && `margin-top: ${getPropertyValue(marginY)};\nmargin-bottom:${getPropertyValue(marginY)};`}
    ${marginLeft && `margin-left: ${getPropertyValue(marginLeft)};`}
    ${marginRight && `margin-right: ${getPropertyValue(marginRight)};`}
    ${marginTop && `margin-top: ${getPropertyValue(marginTop)};`}
    ${marginBottom && `margin-bottom: ${getPropertyValue(marginBottom)};`}

    ${padding && `padding: ${getPropertyValue(padding)}`}
    ${paddingX && `padding-left: ${getPropertyValue(paddingX)};\npadding-right:${getPropertyValue(paddingX)};`}
    ${paddingY && `padding-top: ${getPropertyValue(paddingY)};\npadding-bottom:${getPropertyValue(paddingY)};`}
    ${paddingLeft && `padding-left: ${getPropertyValue(paddingLeft)};`}
    ${paddingRight && `padding-right: ${getPropertyValue(paddingRight)};`}
    ${paddingTop && `padding-top: ${getPropertyValue(paddingTop)};`}
    ${paddingBottom && `padding-bottom: ${getPropertyValue(paddingBottom)};`}
  `

  return style
}

export default spacing
