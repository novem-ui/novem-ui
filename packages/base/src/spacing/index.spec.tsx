import React from 'react'

import { render } from '@testing-library/react'
import styled from '@emotion/styled'

import theme from '@novem-ui/theme'
import spacing from '.'
import type { SpacingProps } from '../types'

const ComponentWithSpacingProps = styled.div<SpacingProps>`
  ${spacing}
`

describe('spacing', () => {
  test('should create a css fragment with the mapped props (partial)', () => {
    const { container } = render(<ComponentWithSpacingProps theme={theme} margin={10} py={4} />)

    expect(container).toMatchSnapshot()
  })

  test('should create a css fragment with the mapped props (including all props)', () => {
    const props: SpacingProps = {
      m: 1,
      mb: '12',
      ml: '1',
      mr: 32,
      mt: '1px',
      mx: 3,
      my: 2,
      p: 2,
      pb: '12',
      pl: '1',
      pr: 32,
      pt: '1px',
      px: 3,
      py: 2
    }

    const { container } = render(<ComponentWithSpacingProps theme={theme} {...props} />)

    expect(container).toMatchSnapshot()
  })
})
