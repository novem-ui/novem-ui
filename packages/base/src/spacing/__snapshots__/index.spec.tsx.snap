// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spacing should create a css fragment with the mapped props (including all props) 1`] = `
.emotion-0 {
  margin: 0.25rem;
  margin-left: 0.75rem;
  margin-right: 0.75rem;
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
  margin-left: 0.25rem;
  margin-right: 8rem;
  margin-top: 1px;
  margin-bottom: 3rem;
  padding:0.5rem padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 0.25rem;
  padding-right: 8rem;
  padding-top: 1px;
  padding-bottom: 3rem;
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;

exports[`spacing should create a css fragment with the mapped props (partial) 1`] = `
.emotion-0 {
  margin: 2.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

<div>
  <div
    class="emotion-0"
  />
</div>
`;
